File: superfields/src/main/resources/META-INF/resources/frontend/clipboard-copy-button.ts
import {html, LitElement} from 'lit';
import {customElement} from 'lit/decorators.js';

@customElement('clipboard-copy-button')
export class ClipboardCopyButton extends LitElement {

    render() {
        return html`<vaadin-icon icon="vaadin:copy" role="img" aria-label="Copy"></vaadin-icon>`;
    }

    setSourceComponent(source: Element) {
        console.info("CCB: setting source to " + source);
    }

    clearSourceComponent() {
        console.info("CCB: clear source component");
    }

}

File: superfields/src/main/java/org/vaadin/miki/superfields/buttons/ClipboardCopyButton.java
package org.vaadin.miki.superfields.buttons;

import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.HasValue;
import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.dependency.JsModule;
import com.vaadin.flow.component.html.Div;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * A button that can always be clicked, no matter the state it is in.
 *
 * @author miki
 * @since 2025-08-16
 */
@Tag("clipboard-copy-button")
@JsModule("./clipboard-copy-button.ts")
public class ClipboardCopyButton extends Div {

  private static final Logger LOGGER = LoggerFactory.getLogger(ClipboardCopyButton.class);

  private HasValue<?, ?> source;

  public HasValue<?, ?> getSource() {
    return source;
  }

  public void setSource(HasValue<?, ?> source) {
    this.source = source;
    LOGGER.info("set source to {}", source);
    if(source != null) {
      this.getElement().getNode().runWhenAttached(ui -> ui.beforeClientResponse(this, context ->
          this.getElement().callJsFunction(
              "setSourceComponent",
              ((Component)this.source).getElement()
          )
      ));
    }
    else this.getElement().getNode().runWhenAttached(ui -> ui.beforeClientResponse(this, context ->
        this.getElement().callJsFunction(
            "clearSourceComponent"
        )
    ));
  }
}
